applications/tari_console_wallet/src/grpc/wallet_grpc_server.rs

```diff
@@ -1032,3 +1076,89 @@ fn convert_wallet_transaction_into_transaction_info(
         },
     }
 }
+
+const EVENT_INDEX: usize = 0;
+const AMOUNT_INDEX: usize = 1;
+const TX_ID_INDEX: usize = 2;
+const MESSAGE_INDEX: usize = 3;
+const PK_INDEX: usize = 4;
+const INCOMPLETED_STATUS_INDEX: usize = 5;
+const INCOMPLETED_DIRECTION_INDEX: usize = 6;
+
+const COMPLETED_PUBLIC_KEY_INDEX: usize = 4;
+const COMPLETED_SOURCE_KEY_INDEX: usize = 5;
+const COMPLETED_STATUS_INDEX: usize = 6;
+const COMPLETED_DIRECTION_INDEX: usize = 11;
+
+fn convert_vec_into_transaction_event(source: Vec<String>) -> Option<TransactionEvent> {
+    if source.len() != 7 || source.len() != 12 {
+        error!(
+            "The source must have exactly 7 or 12 elements. The current len: {}",
+            source.len()
+        );
+        return None;
+    }
+
+    let event = source.iter().nth(EVENT_INDEX).unwrap().clone();
+    let amount = match source.iter().nth(AMOUNT_INDEX).unwrap().clone().parse::<u64>() {
+        Ok(amt) => amt,
+        Err(_) => return None,
+    };
+    let tx_id = match source.iter().nth(TX_ID_INDEX).unwrap().clone().parse::<u64>() {
+        Ok(id) => id,
+        Err(_) => return None,
+    };
+    let message = source.iter().nth(MESSAGE_INDEX).unwrap().clone();
+
+    let mut event = TransactionEvent {
+        event,
+        amount,
+        tx_id,
+        message,
+        ..Default::default()
+    };
+
+    if source.len() == 7 {
+        let pk = source.iter().nth(PK_INDEX).unwrap().clone().into_bytes();
+        let status = source.iter().nth(INCOMPLETED_STATUS_INDEX).unwrap().clone();
+        let direction = source.iter().nth(INCOMPLETED_DIRECTION_INDEX).unwrap().clone();
+
+        event.status = status;
+        event.direction = direction.clone();
+
+        match direction.to_lowercase().as_str() {
+            "inbound" => {
+                event.dest_pk = pk.clone();
+            },
+            "outbound" => {
+                event.source_pk = pk.clone();
+            },
+            _ => (),
+        }
+        Some(event)
+    } else if source.len() == 12 {
+        let source_pk = source
+            .iter()
+            .nth(COMPLETED_SOURCE_KEY_INDEX)
+            .unwrap()
+            .clone()
+            .into_bytes();
+        let dest_pk = source
+            .iter()
+            .nth(COMPLETED_PUBLIC_KEY_INDEX)
+            .unwrap()
+            .clone()
+            .into_bytes();
+        let status = source.iter().nth(COMPLETED_STATUS_INDEX).unwrap().clone();
+        let direction = source.iter().nth(COMPLETED_DIRECTION_INDEX).unwrap().clone();
+
+        event.dest_pk = dest_pk;
+        event.source_pk = source_pk;
+        event.status = status;
+        event.direction = direction;
+
+        Some(event)
+    } else {
+        Option::None
+    }
```

Comment:
It would be more readable to have some helper functions here:

```rust
match source.len() {
   7 => extract_foo_args(),
  12 => extract_baz_args(),
}
```